### ISSUE TITLE
Without flash?

### ISSUE THREAD

**Issue by kaatish** [kbhomes/radiant-player-mac#335]
Hi,
Is there a possibility to use radiant player without flash support? Is this planned sometime in the future?

**Comment by izacus** [comment]
Did you enable the experimental HTML5 audio option in the Google Play Music labs?

**Comment by ccampbell** [comment]
I don't know what changed, but starting today radiant player is asking me to install flash even though I have the html5 option enabled.  Looks like google dropped HTML5 audio support :cry:

![](https://devi.vimeocdn.com/craig/AUTm)

**Comment by repinel** [comment]
I wound't say the dropped the support. You can see by the settings that it may not always work

>HTML5 Audio
>Listen to your music without the need for Flash. Support for this lab is still experimental, and it may not work in all cases.

**Comment by ccampbell** [comment]
@repinel good point.  There is one interesting thing which is that in chrome with flash disabled it still works though and is using HTML5 audio.  When I open Safari it says it requires flash.

![8rjf](https://cloud.githubusercontent.com/assets/259316/8312560/4133d802-19ab-11e5-941c-ff0b3e76b88b.png)

**Comment by lacymorrow** [comment]
I got this pop up today as well

**Comment by lnzero1** [comment]
Please make it possible to use it without flash - does anybody know which restrictions google apply to their code to prevent playing without flash player?
I tried the following (on OS X 10.10):
Play music opened with Chrome 45 ✓
Play music opened with Opera Next ❌
Play music opened with Opera Next, User agent set to be identical to Chrome 45 ❌
Play music opened with Safari ❌

I red that opera next and safari, both, should support HTML5 audio API?

Due to the current security breaches within flash player, i completely removed flash player from all my current systems and I'm not going to reinstall that crappy part of software.

So last chance for radiant player to get my attention, if no-flash-requirement is fullfilled.
Otherwise ⌘-Backspace will do the thing.

Thanks for your developing time anyway.

**Comment by ccampbell** [comment]
I was doing some research, and I that Google switched to using mediasource to playback the audio.  

https://developers.google.com/web/updates/2015/06/mse-gapless-audio?hl=en

That would explain why it is only available in the latest chrome

**Comment by kaatish** [comment]
Hi, Yes as others have pointed out, I have enabled HTML5 option, however the app still asks for missing flash plugin while chrome does not.

**Comment by bondt** [comment]
Same here; HTML5 Audio support enabled, but still a popup and no music.

**Comment by kbhomes** [comment]
This support is actually planned in the future. The HTML5 audio support in Google Play Music has been closed off by Google to specifically the Chrome platform, and rightly so because it seems that it was the only platform, by their testing I'm sure, that supported the HTML5 Media Source Extensions. This meant that Safari, Firefox, and other browsers would never use the HTML5 audio.

Good news though! Last couple weeks, I've worked on attempting to get this working in WebKit, by spoofing the Chrome user agent and seeing how WebKit does/doesn't fare with playing the audio without Flash. I dug into the WebKit source code, found the issue that was preventing audio playback, and [filed a bug and a patch for WebKit that landed last week][1].

What this means is that with a little user agent spoofing (and a JS workaround for [this WebKit bug I've also reported][2] until that gets fixed), HTML5 audio playback should work in Safari and Radiant Player on OS X 10.11. 

Now, for OS X 10.9 and 10.10, it may be the case that the system's WebKit.framework does not get updated to a version that includes my patch. In that case, it may be necessary to download the Nightly release of WebKit and use some Terminal commands to get Radiant Player working with HTML5 audio, but that may be complicated and I will hold off on discussing that until I get more information on the state of WebKit in non-El Capitan versions of OS X.

Perhaps Google may even eventually open up the HTML5 audio support on the site to Chrome *and* Safari, which would be a huge win for all users hoping to leave Flash.

[1]: https://bugs.webkit.org/show_bug.cgi?id=147252
[2]: https://bugs.webkit.org/show_bug.cgi?id=147309

**Comment by mendelk** [comment]
:+1:

**Comment by repinel** [comment]
@kbhomes Thank you for working this :smile:

**Comment by t3rminus** [comment]
What about using something like chromiumembedded (https://bitbucket.org/chromiumembedded/), instead of WebKit.framework?

**Comment by stevenla** [comment]
with stuff [like this happening on a monthly basis](https://helpx.adobe.com/security/products/flash-player/apsa15-04.html), adding proper html5 support would be amazing.

**Comment by bcardiff** [comment]
:+1:

**Comment by jacobwgillespie** [comment]
It's worth noting that running Google Play Music inside an Electron app doesn't require a separate installation of flash.  I believe it's using HTML5 to do the streaming (since it's a Chrome environment) though I'm not 100% sure on that (it does work without explicitly allowing plugins, which I believe means that Flash is not running).

There's [radiant-player-electron](https://github.com/radiant-player/radiant-player-electron) if you're curious, though it's definitely a test project and nowhere near feature complete.

**Comment by t3rminus** [comment]
Although nowhere near as feature-complete as Radiant Player, there's also this: https://github.com/twolfson/google-music-electron, which shows its definitely possible, and works well without flash.

**Comment by jrajav** [comment]
@kbhomes Any chance you can go into a little more detail on how to perform that user agent spoofing and JS workaround? I tried simply setting the user agent to Chrome and I still get "You must install Flash player" on latest Safari in El Capitan.

**Comment by MarshallOfSound** [comment]
@jrajav It won't just be spoofing the useragent.  If you trace the GPM core js with a debugger you will see it checks for AudioContext functioanlity.

Safari does not yet fully implement the WebAudio API so it will not play the music through HTML5 audio.

See here --> http://caniuse.com/#feat=audio-api

**Comment by Axiol** [comment]
One solution could be to bundle Flash inside Radiant. Like Chrome does. As long as I don't have to install Flash on my system, I'm fine with that :)

**Comment by lacymorrow** [comment]
@Axiol I like that idea, although I'm not sure how adobe would feel about bundling a flash binary.

**Comment by jacobwgillespie** [comment]
Yeah, this has been tried in Radiant in the past and just ends up with an out-of-date copy of Flash embedded inside (which isn't fun for security).

We're working on Radiant 2.0 which will support HTML5 audio and remove the Flash requirement.

**Comment by zasadnyy** [comment]
Any updates on this one?

**Comment by ntimo** [comment]
Any news on this?

**Comment by chrismou** [comment]
Hey guys - still on the horizon, but a bit delayed due to real world commitments.

Fingers crossed we should have a stable (ish) release ready to play with in the near future.

**Comment by ccampbell** [comment]
For anyone curious there is another desktop app for Google music that does not require flash so I have been using that until radiant player is updated:

https://www.googleplaymusicdesktopplayer.com

**Comment by joosebox** [comment]
Thanks for the link ccampbell! This app fills the void of no Radiant Player.
